vsphere/internal/virtualdevice/virtual_machine_cdrom_subresource.go

```diff
@@ -0,0 +1,507 @@
+	log.Printf("[DEBUG] CdromPostCloneOperation: Device config operations from post-clone: %s", DeviceChangeString(spec))
+	log.Printf("[DEBUG] CdromPostCloneOperation: Operation complete, returning updated spec")
+	return l, spec, nil
+}
+
+// Create creates a vsphere_virtual_machine cdrom sub-resource.
+func (r *CdromSubresource) Create(l object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	log.Printf("[DEBUG] %s: Running create", r)
+	var spec []types.BaseVirtualDeviceConfigSpec
+	var ctlr types.BaseVirtualController
+	ctlr, err := r.ControllerForCreateUpdate(l, SubresourceControllerTypeIDE, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// We now have the controller on which we can create our device on.
+	dsID := r.Get("datastore_id").(string)
+	path := r.Get("path").(string)
+	ds, err := datastore.FromID(r.client, dsID)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find datastore: %s", err)
+	}
+	dsProps, err := datastore.Properties(ds)
+	if err != nil {
+		return nil, fmt.Errorf("could not get properties for datastore: %s", err)
+	}
+	dsName := dsProps.Name
+
+	dsPath := &object.DatastorePath{
+		Datastore: dsName,
+		Path:      path,
+	}
+
+	device, err := l.CreateCdrom(ctlr.(*types.VirtualIDEController))
+	if err != nil {
+		return nil, err
+	}
+	device = l.InsertIso(device, dsPath.String())
+
+	// Done here. Save IDs, push the device to the new device list and return.
+	r.SaveDevIDs(device, ctlr)
+	dspec, err := object.VirtualDeviceList{device}.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+	if err != nil {
+		return nil, err
+	}
+	spec = append(spec, dspec...)
+	log.Printf("[DEBUG] %s: Device config operations from create: %s", r, DeviceChangeString(spec))
+	log.Printf("[DEBUG] %s: Create finished", r)
+	return spec, nil
+}
+
+// Read reads a vsphere_virtual_machine cdrom sub-resource.
+func (r *CdromSubresource) Read(l object.VirtualDeviceList) error {
+	log.Printf("[DEBUG] %s: Reading state", r)
+	d, err := r.FindVirtualDevice(l)
+	if err != nil {
+		return fmt.Errorf("cannot find disk device: %s", err)
+	}
+	device, ok := d.(*types.VirtualCdrom)
+	if !ok {
+		return fmt.Errorf("device at %q is not a virtual CDROM device", l.Name(d))
+	}
+	if backing, ok := device.Backing.(*types.VirtualCdromIsoBackingInfo); ok {
+		// Only read backing info if it's available. If not, this is a host
+		// device/client mapping that is not managed by TF and needs to be deleted.
+		dp := &object.DatastorePath{}
+		if ok := dp.FromString(backing.FileName); !ok {
+			return fmt.Errorf("could not read datastore path in backing %q", backing.FileName)
+		}
+		r.Set("datastore_id", backing.Datastore.Value)
+		r.Set("path", dp.Path)
+	}
+	// Save the device key and address data
+	ctlr, err := findControllerForDevice(l, d)
+	if err != nil {
+		return err
+	}
+	r.SaveDevIDs(d, ctlr)
+	log.Printf("[DEBUG] %s: Read finished (key and device address may have changed)", r)
```

Comment:
Again, this will show the sub-resource address - all sub-resources implement Stringer, which will print something like `cdrom.0`, `network_interface.0`, `disk.0`, etc.